dpnp/dpnp_iface_linearalgebra.py

```diff
@@ -269,63 +285,364 @@ def matmul(x1, x2, out=None, **kwargs):
+            x1 = x1[dpnp.newaxis, :]
+            x1_ndim = x1.ndim
+
+        if x2_ndim == 1:
+            x2 = x2[:, dpnp.newaxis]
+            x2_ndim = x2.ndim
+
+        x1_shape = x1.shape
+        x2_shape = x2.shape
+        if x1_shape[-1] != x2_shape[-2]:
+            raise ValueError(
+                "Input operand 1 has a mismatch in its core dimension 0, "
+                "with gufunc signature (n?,k),(k,m?)->(n?,m?) "
+                f"(size {x1_shape[1]} is different from {x2_shape[0]})"
+            )
+
+        # Determine the appropriate data types
+        if dtype is not None:
+            res_dtype = dtype
+            gemm_dtype, _ = _gemm_res_dtype(x1, x2, casting=casting)
+        else:
+            gemm_dtype, res_dtype = _gemm_res_dtype(x1, x2, casting=casting)
+        if x1.dtype != gemm_dtype:
+            x1 = dpnp.astype(x1, gemm_dtype, casting=casting)
+        if x2.dtype != gemm_dtype:
+            x2 = dpnp.astype(x2, gemm_dtype, casting=casting)
+
+        # find the result shape
+        x1_is_2D = False
+        x1_is_2D = False
+        if x1_ndim == 2 and x2_ndim == 2:
+            res_shape = (x1.shape[0], x2.shape[1])
+        else:
+            x1_is_2D = dpnp.all(dpnp.array(x1_shape[:-2]) == 1)  # inherently 2D
+            x2_is_2D = dpnp.all(dpnp.array(x2_shape[:-2]) == 1)
+
+            # makes the dimension of input the same by adding new axis
+            if x1_ndim != x2_ndim:
+                diff = abs(x1_ndim - x2_ndim)
+                if x1_ndim < x2_ndim:
+                    x1 = x1.reshape((1,) * diff + x1.shape)
+                    x1_ndim = x1.ndim
+                    x1_shape = x1.shape
+                else:
+                    x2 = x2.reshape((1,) * diff + x2.shape)
+                    x2_ndim = x2.ndim
+                    x2_shape = x2.shape
+
+            # examining the option to align inputs when their shapes differ but they are 1-D in some dimensions.
+            tmp_shape = list(x1_shape[:-2])
+            for i in range(x1_ndim - 2):
+                if x1_shape[i] != x2_shape[i]:
+                    if x1_shape[i] == 1:
+                        tmp_shape[i] = x2_shape[i]
+                        # If the `x1` array is inherently 2D, there's no need to duplicate the data for the 1-D dimension; GEMM handles it automatically.
+                        if not x1_is_2D:
+                            x1 = dpnp.repeat(x1, x2_shape[i], axis=i)
+                    elif x2_shape[i] == 1:
+                        tmp_shape[i] = x1_shape[i]
+                        if not x2_is_2D:
+                            x2 = dpnp.repeat(x2, x1_shape[i], axis=i)
+                    else:
+                        raise ValueError(
+                            "operands could not be broadcast together with remapped shapes."
+                        )
+            x1_shape = x1.shape
+            x2_shape = x2.shape
+            res_shape = tuple(tmp_shape) + (x1_shape[-2], x2_shape[-1])
+
```

Comment:
[Tag: copy]
This will avoid multiple copies and improve performance
Also created a list of events to achieve parallel copying for `x2` array with copying `x1` array 
```suggestion
        is_x1_c_f_contig = x1.flags.c_contiguous or x1.flags.f_contiguous
        is_x2_c_f_contig = x2.flags.c_contiguous or x2.flags.f_contiguous

        list_of_events = []

        # input arrays should be C_CONTIGUOUS or F_CONTIGUOUS
        if not is_x1_c_f_contig or x1.dtype != gemm_dtype:
            x1_copy = dpnp.empty_like(x1, dtype=gemm_dtype, order="C")
            ht_copy_ev_x1, _ = ti._copy_usm_ndarray_into_usm_ndarray(
                src=dpnp.get_usm_ndarray(x1),
                dst=x1_copy.get_array(),
                sycl_queue=x1.sycl_queue,
            )
            x1 = x1_copy
            list_of_events.append(ht_copy_ev_x1)

        if not is_x2_c_f_contig or x2.dtype != gemm_dtype:
            x2_copy = dpnp.empty_like(x2, dtype=gemm_dtype, order="C")
            ht_copy_ev_x2, _ = ti._copy_usm_ndarray_into_usm_ndarray(
                src=dpnp.get_usm_ndarray(x2),
                dst=x2_copy.get_array(),
                sycl_queue=x2.sycl_queue,
            )
            x2 = x2_copy
            list_of_events.append(ht_copy_ev_x2)

        if list_of_events:
            for ev in list_of_events:
                ev.wait()
```